### PR TITLE
DBCON-329, DBCON-351, DBCON-352: Change how createARRAY and createArrayOf are exposed, use unwrap to get getARRAY, Oracle features in Spring originated connections

### PR DESCRIPTION
**Pull request by niqueco**
This PR:

- Makes the handling of createArrayOf and createARRAY explicit and clear.
- Uses unwrap to access Oracle's createARRAY(), which could potentially play well with connection pool implementations.
- Allow accessing Oracle features in Spring originated connections.
- Removes the wrapper used for Oracle connections.

### CODE REVIEW

File: src/main/java/org/mule/extension/db/internal/domain/connection/datasource/DbDataSourceReferenceConnectionProvider.java [hunk 1]
```diff
@@ -29,6 +36,15 @@
     nameRegexpMatcher = DRIVER_FILE_NAME_PATTERN, type = JAR, optional = true)
 public class DbDataSourceReferenceConnectionProvider extends DataSourceReferenceConnectionProvider {
 
+  private final Map<String, Map<Integer, ResolvedDbType>> resolvedDbTypesCache = new ConcurrentHashMap<>();
 
+  @Override
+  protected DbConnection createDbConnection(Connection connection) throws Exception {
+    if (connection.getMetaData().getDatabaseProductName().equals("Oracle")) {
```
Comment (nicomz) [inline src/main/java/org/mule/extension/db/internal/domain/connection/datasource/DbDataSourceReferenceConnectionProvider.java:43]: Should we introduce a new Helper class to distinguish vendors given a connection?

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleDbConnection.java [hunk 2]
```diff
@@ -72,8 +73,6 @@
   private static final int PROCEDURE_NAME = 3;
   private static final int PARAM_NAME_COLUMN_INDEX = 4;
 
-  private Method createArrayMethod;
-
   Map<String, Map<Integer, ResolvedDbType>> resolvedDbTypesCache;
```
Comment (nicomz) [inline src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleDbConnection.java:76]: Should this be final? (resolvedDbTypesCache)

**Review by nicomz — CHANGES_REQUESTED** [review]

File: src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleDbConnection.java (hunk 2 above)
Comment (niqueco) [reply src/main/java/org/mule/extension/db/internal/domain/connection/oracle/OracleDbConnection.java:76]: Yes, totally, and private as well.

File: src/main/java/org/mule/extension/db/internal/domain/connection/datasource/DbDataSourceReferenceConnectionProvider.java (hunk 1 above)
Comment (niqueco) [reply src/main/java/org/mule/extension/db/internal/domain/connection/datasource/DbDataSourceReferenceConnectionProvider.java:43]: Yes, that's a good idea. What if I add a `handles` static method to `OracleDbConnection`?

The code would then be...

```
    if (OracleDbConnection.handles(connection)) {
      return new OracleDbConnection(connection, super.resolveCustomTypes(), resolvedDbTypesCache);
    } else {
      return super.createDbConnection(connection);
    }
```

**Review by juldiazmule — DISMISSED** [review]

**Review by nicomz — CHANGES_REQUESTED** [review]

**Review by nicomz — APPROVED** [review]

**Pull request merged**